src/container/srv_container.c

```diff
@@ -960,6 +971,308 @@ cont_destroy(struct rdb_tx *tx, struct ds_pool_hdl *pool_hdl,
 	return rc;
 }
 
+struct cont_ec_agg *
+cont_ec_agg_lookup(struct cont_svc *cont_svc, uuid_t cont_uuid)
+{
+	struct cont_ec_agg *ec_agg;
+
+	d_list_for_each_entry(ec_agg, &cont_svc->cs_ec_agg_list, ea_list) {
+		if (uuid_compare(ec_agg->ea_cont_uuid, cont_uuid) == 0)
+			return ec_agg;
+	}
+	return NULL;
+}
+
+static int
+cont_ec_agg_alloc(struct cont_svc *cont_svc, uuid_t cont_uuid,
+		  struct cont_ec_agg **ec_aggp)
+{
+	struct cont_ec_agg	*ec_agg = NULL;
+	struct pool_domain	*doms;
+	int			node_nr;
+	int			rc = 0;
+	int			i;
+
+	D_ALLOC_PTR(ec_agg);
+	if (ec_agg == NULL)
+		return -DER_NOMEM;
+
+	D_ASSERT(cont_svc->cs_pool->sp_map != NULL);
+	node_nr = pool_map_find_nodes(cont_svc->cs_pool->sp_map,
+				      PO_COMP_ID_ALL, &doms);
+	if (node_nr < 0)
+		D_GOTO(out, rc = node_nr);
+
+	D_ALLOC_ARRAY(ec_agg->ea_server_ephs, node_nr);
+	if (ec_agg->ea_server_ephs == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	uuid_copy(ec_agg->ea_cont_uuid, cont_uuid);
+	ec_agg->ea_servers_num = node_nr;
+	ec_agg->ea_current_eph = 0;
+	for (i = 0; i < node_nr; i++) {
+		ec_agg->ea_server_ephs[i].rank = doms[i].do_comp.co_rank;
+		ec_agg->ea_server_ephs[i].eph = 0;
+	}
+	d_list_add(&ec_agg->ea_list, &cont_svc->cs_ec_agg_list);
+	*ec_aggp = ec_agg;
+out:
+	if (rc) {
+		if (ec_agg && ec_agg->ea_server_ephs)
+			D_FREE(ec_agg->ea_server_ephs);
+		if (ec_agg)
+			D_FREE(ec_agg);
+	}
+
+	return rc;
+}
+
+static void
+cont_ec_agg_destroy(struct cont_ec_agg *ec_agg)
+{
+	d_list_del(&ec_agg->ea_list);
+	if (ec_agg->ea_server_ephs)
+		D_FREE(ec_agg->ea_server_ephs);
+	D_FREE(ec_agg);
+}
+
+/**
+ * Update the epoch (by rank) on the leader of container service, which
+ * will be called by IV update on the leader.
+ */
+int
+ds_cont_leader_update_agg_eph(uuid_t pool_uuid, uuid_t cont_uuid,
+			      d_rank_t rank, daos_epoch_t eph)
+{
+	struct cont_svc		*svc;
+	struct cont_ec_agg	*ec_agg;
+	int			rc;
+	bool			retried = false;
+	int			i;
+
+	rc = cont_svc_lookup_leader(pool_uuid, 0 /* id */, &svc,
+				    NULL /* hint */);
+	if (rc != 0)
+		return rc;
+
+retry:
+	ec_agg = cont_ec_agg_lookup(svc, cont_uuid);
+	if (ec_agg == NULL) {
+		rc = cont_ec_agg_alloc(svc, cont_uuid, &ec_agg);
+		if (rc)
+			D_GOTO(out_put, rc);
+	}
+
+	for (i = 0; i < ec_agg->ea_servers_num; i++) {
+		if (ec_agg->ea_server_ephs[i].rank == rank) {
+			if (ec_agg->ea_server_ephs[i].eph < eph)
+				ec_agg->ea_server_ephs[i].eph = eph;
+			break;
+		}
+	}
+
+	if (i == ec_agg->ea_servers_num) {
```

Comment:
Why we do realloc until some rank failed to find entry in this 'ec_agg'? That will invalidate all other valid reported epochs.

Shouldn't we always check pool map first, then do realloc immediately once "ea_servers_num != node_nr"?

Reply:
I think realloc will keep the original data. 
Well,  I want to avoid pool map checking every time